provider/equinix/environ.go

```diff
@@ -0,0 +1,777 @@
+	}
+
+	arch := getArchitectureFromPlan(d.Plan.Name)
+
+	return &environs.StartInstanceResult{
+		Instance: inst,
+		Hardware: &instance.HardwareCharacteristics{
+			Arch: &arch,
+			Mem:  &mem,
+			// RootDisk: &instanceSpec.InstanceType.RootDisk,
+			CpuCores: &cpus,
+		},
+	}, nil
+}
+
+func (e *environ) StopInstances(ctx context.ProviderCallContext, ids ...instance.Id) error {
+	return e.deleteDevicesByIDs(ctx, ids)
+}
+
+func (e *environ) StorageProvider(t storage.ProviderType) (storage.Provider, error) {
+	return nil, errors.NotFoundf("storage provider %q", t)
+}
+
+func (e *environ) StorageProviderTypes() ([]storage.ProviderType, error) {
+	return nil, nil
+}
+
+// Region is specified in the HasRegion interface.
+func (e *environ) Region() (simplestreams.CloudSpec, error) {
+	return simplestreams.CloudSpec{
+		Region:   e.cloud.Region,
+		Endpoint: e.cloud.Endpoint,
+	}, nil
+}
+
+func (e *environ) deleteDevicesByIDs(ctx context.ProviderCallContext, ids []instance.Id) error {
+	for _, id := range ids {
+		_, err := e.equinixClient.Devices.Delete(string(id), true)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+// if values tag and state are left empty it will return all instances
+func (e *environ) getPacketInstancesByTag(tags map[string]string) ([]instances.Instance, error) {
+	var toReturn []instances.Instance
+
+	deviceTags := set.NewStrings()
+	for k, v := range tags {
+		deviceTags.Add(fmt.Sprintf("%s=%s", k, v))
+	}
+
+	projectID, ok := e.cloud.Credential.Attributes()["project-id"]
+	if !ok {
+		return nil, fmt.Errorf("project-id not fond as attribute")
+	}
+
+	devices, _, err := e.equinixClient.Devices.List(projectID, nil)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	for _, d := range devices {
+		cp := d
+		cpTags := set.NewStrings(cp.Tags...)
+		if !deviceTags.Intersection(cpTags).IsEmpty() {
+			toReturn = append(toReturn, &equinixDevice{e, &cp})
+		}
+	}
+
+	return toReturn, nil
+}
+
+func (e *environ) getSubnetsToZoneMap(ctx context.ProviderCallContext, args environs.StartInstanceParams) ([]string, error) {
+	var subnetIDs []string
+	for _, subnetList := range args.SubnetsToZones {
+		for subnetID := range subnetList {
+			packetSubnetID := strings.TrimPrefix(subnetID.String(), "subnet-")
+			subnetIDs = append(subnetIDs, packetSubnetID)
+		}
+	}
+
+	return subnetIDs, nil
+}
+
+// supportedInstanceTypes returns the instance types supported by Equnix Metal.
+func (e *environ) supportedInstanceTypes() ([]instances.InstanceType, error) {
+	opt := &packngo.ListOptions{
+		Includes: []string{"available_in_metros"},
+	}
+	plans, _, err := e.equinixClient.Plans.List(opt)
+	if err != nil {
+		return nil, errors.Annotate(err, "retrieving supported instance types")
+	}
+
+	var instTypes []instances.InstanceType
+nextPlan:
+	for _, plan := range plans {
+		if !validPlan(plan, e.cloud.Region) {
+			logger.Debugf("Plan %s not valid in facility %s", plan.Name, e.cloud.Region)
```

Comment:
@gianarb Can we reduce the logging level on this, it's very chatty.

Reply:
I can only remove that I suppose. Are you suggesting so?

Reply:
Can we drop it to `Tracef`.